javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -2,167 +2,187 @@ package java.util
+    sortRangeImpl[Float](a, fromIndex, toIndex)
 
-  def fill(a: Array[Float], value: Float): Unit =
-    fillImpl(a, value)
+  @noinline def sort(a: Array[Double]): Unit =
+    sortImpl(a)
 
-  def fill(a: Array[Float], fromIndex: Int, toIndex: Int, value: Float): Unit =
-    fillImpl(a, fromIndex, toIndex, value)
+  @noinline def sort(a: Array[Double], fromIndex: Int, toIndex: Int): Unit =
+    sortRangeImpl[Double](a, fromIndex, toIndex)
 
-  def fill(a: Array[Double], value: Double): Unit =
-    fillImpl(a, value)
+  @noinline def sort(a: Array[AnyRef]): Unit =
+    sortImpl(a)
 
-  def fill(a: Array[Double], fromIndex: Int, toIndex: Int, value: Double): Unit =
-    fillImpl(a, fromIndex, toIndex, value)
+  @noinline def sort(a: Array[AnyRef], fromIndex: Int, toIndex: Int): Unit =
+    sortRangeImpl[AnyRef](a, fromIndex, toIndex)
 
-  private def fillImpl[@specialized T](a: Array[T], value: T): Unit = {
-    var i = 0
-    while (i != a.length) {
-      a(i) = value
-      i += 1
-    }
+  @noinline def sort[T <: AnyRef](array: Array[T], comparator: Comparator[_ >: T]): Unit = {
+    implicit val ord = toOrdering(comparator).asInstanceOf[Ordering[AnyRef]]
+    sortImpl(array.asInstanceOf[Array[AnyRef]])
   }
 
-  private def fillImpl[@specialized T](a: Array[T],
-      fromIndex: Int, toIndex: Int, value: T): Unit = {
-    if (fromIndex > toIndex)
-      throw new IllegalArgumentException
-    if (fromIndex < 0 || toIndex > a.length)
-      throw new ArrayIndexOutOfBoundsException
-
-    var i = fromIndex
-    while (i != toIndex) {
-      a(i) = value
-      i += 1
-    }
+  @noinline def sort[T <: AnyRef](array: Array[T], fromIndex: Int, toIndex: Int,
+      comparator: Comparator[_ >: T]): Unit = {
+    implicit val ord = toOrdering(comparator).asInstanceOf[Ordering[AnyRef]]
+    sortRangeImpl[AnyRef](array.asInstanceOf[Array[AnyRef]], fromIndex, toIndex)
   }
 
-  def fill(a: Array[AnyRef], value: AnyRef): Unit = {
-    var i = 0
-    while (i < a.length) {
-      a(i) = value
-      i += 1
-    }
+  @inline
+  private def sortRangeImpl[@specialized T: ClassTag](a: Array[T], fromIndex: Int,
+      toIndex: Int)(implicit ord: Ordering[T]): Unit = {
+    checkIndicesForCopyOfRange(a.length, fromIndex, toIndex)
+    stableSortRangeImpl(a, fromIndex, toIndex - 1, new Array[T](a.length), ord.lt)
   }
 
-  def fill(a: Array[AnyRef],
-      fromIndex: Int, toIndex: Int, value: AnyRef): Unit = {
-    if (fromIndex > toIndex)
-      throw new IllegalArgumentException
-    if (fromIndex < 0 || toIndex > a.length)
-      throw new ArrayIndexOutOfBoundsException
-
-    var i = fromIndex
-    while (i < toIndex) {
-      a(i) = value
-      i += 1
+  @inline
+  private def sortImpl[T: ClassTag](a: Array[T])(implicit ord: Ordering[T]): Unit =
+    stableSortRangeImpl(a, 0, a.length - 1, new Array[T](a.length), ord.lt)
+
+  private def stableSortRangeImpl[T](a: Array[T], lo: Int, hi: Int,
+      scratch: Array[T], lt: (T, T) => Boolean): Unit = {
```

Comment:
Actually, ideally there should be a manual copy for `T <: AnyRef`, named something like `stableSortRangeRefImpl` and called from `sort` overloads taking an array of references. This is because, unfortunately, `@specialized` does not specialize for the `AnyRef` case.